fmt-c-writer.scm

```diff
@@ -52,77 +53,214 @@
+     (make-field-access form))
+    (('var . _) (walk-var form))
+    (('cast expr type) (list '%cast
+                             (walk-type type)
+                             (walk-expr expr)))
+    (('enum . _) (walk-enum form))
+    ;; | is problematic... And c-or/bit-or/etc are actually
+    ;; procedures, so we have to call the procedure itself
+    (('c-or . rest) (apply c-or (map walk-expr rest)))
+    (('c-bit-or . rest) (apply c-bit-or (map walk-expr rest)))
+    (('c-bit-or= . rest) (apply c-bit-or= (map walk-expr rest)))
+    (else (map walk-expr form))))
+
+(define (walk-var form)
+  ;; (var a int) -> (%var int a)
+  ;; (var a (const int) 32) -> (%var (const int) a 32)
+  ;; (var b [const char 512]) -> (%var (%array (const char) 512) b)
+  ;; note: [...] is actually (¤ ...) after reading
+  ;; (var c (fn ((int) (float)) void)) -> (%var (%fun void ((int) (float))) c)
+  `(%var
+    ,(walk-type (third form))
+    ,(atom-to-fmt-c (second form))
+    .
+    ,(if (null? (drop form 3))
+         (list)
+         (walk-expr (drop form 3)))      ; optional init expression
+    ))
+
+(define (walk-type form)
+  ;; int -> int
+  ;; (const int) -> const int
+  ;; [const char 512] ->(¤ (const char) 512) -> (%array (const char) 512)
+  ;; [float 8] -> (%array float 8)
+  ;; (* const char) -> (const char *)
+  ;; (const * const * const char) -> (const char * const * const)
+  ;; (fn ((int) (float)) void) -> (%fun void ((int) (float)))
+  (match form
+    (('¤ . array-type)
+     (if (integer? (last array-type))
+         ;; sized array
+         (let* ((type-list (drop-right array-type 1))
+                (type (if (and (list? (car type-list))
+                               (= 1 (length type-list)))
+                          (car type-list)
+                          type-list))
+                (size (last array-type)))
+           `(%array ,(walk-type type)
+                    ,size))
+         ;; sugar for pointer... Do we really need it? Guess why not,
+         ;; it's a strong semantic cue
+         `(%array ,(walk-type (if (and (list? (car array-type))
+                                       (= 1 (length array-type)))
+                                  (car array-type)
+                                  array-type)))))
+    (('fn arglist ret-type)
+     `(%fun ,(walk-type ret-type) ,(walk-arglist arglist)))
+
+    ;; Special case: nested structs/unions
+    ((or ('struct . _)
+         ('union . _)) (walk-struct form))
+
+    (('enum . _) (walk-enum form))
+    (else
+     (type-convert-to-c form))))
+
+(define (type-convert-to-c type)
+  ;; Our pointers to C pointers
+  ;; int -> int
+  ;; * const char -> const char *
+  ;; const * const char -> const char * const
+  (if (atom? type) (atom-to-fmt-c type)
+      (flatten
+       (tree-map atom-to-fmt-c
+                 (fold-right append (list)
+                             (list-join (reverse (list-split type '*))
+                                        '(*)))))))
```

Comment:
That function is a little complex, would be nice to have unit tests for it.

Also could fold-right be replaced with flatten?
```suggestion
  (if (atom? type) (atom-to-fmt-c type)
      (flatten
       (tree-map atom-to-fmt-c
                 (flatten
                             (list-join (reverse (list-split type '*))
                                        '(*)))))))
```

Gosh it's fucking torture to write and edit lisp snippets on github.

Reply:
Yes. Qualifies as self-harm.

Reply:
>Also could fold-right be replaced with flatten?

In general, it could not:

```
(fold-right append (list) '(((1 2)) (3 4) (5 6)))
; ((1 2) 3 4 5 6)
```

However, here we can, I think. There shouldn't be any meaningful nesting in `cv*` type chains. 

> That function is a little complex, would be nice to have unit tests for it.

It's implicitly tested via all `walk-type` tests. However, I agree that it would be nice to have some test specifically for this function.